### PR TITLE
gh-95388: Deprecate creating immutable types with mutable bases

### PR DESCRIPTION
**Pull request by encukou**
* Issue: gh-95388

### CODE REVIEW

File: Lib/test/test_capi.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 import time
 import unittest
 import weakref
+import warnings
```
Comment (erlend-aasland) [inline Lib/test/test_capi.py:18]: Nit: insert it before `import weakref` so the list remains sorted :)

File: Lib/test/test_capi.py [hunk 2]
```diff
@@ -625,6 +626,32 @@ def test_pytype_fromspec_with_repeated_slots(self):
...
+
+        with self.assertWarns(DeprecationWarning):
+            ImmutableSubclass = _testcapi.make_immutable_type_with_base(
+                MutableBase)
+        instance = ImmutableSubclass()
+
+        self.assertEqual(instance.meth(), 'original')
+
+        # Cannot override the static type's method
+        with self.assertRaises(TypeError):
```
Comment (erlend-aasland) [inline Lib/test/test_capi.py]: I'd narrow it down with `assertRaisesRegex`.

File: Objects/typeobject.c [hunk 3]
```diff
@@ -3666,6 +3666,32 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
         goto finally;
     }
 
+    /* If this is an immutable type, check if all bases are also immutable,
+     * and (for now) fire a deprecation warning if not.
+     * (This isn't necessary for static types: those can't have heap bases,
+     * and only heap types can be mutable.)
+     */
+    if (spec->flags & Py_TPFLAGS_IMMUTABLETYPE) {
```
Comment (erlend-aasland) [inline Objects/typeobject.c:3674]:
```suggestion
    if (spec->flags & Py_TPFLAGS_IMMUTABLETYPE &&
        spec->flags & Py_TPFLAGS_HEAPTYPE)
    {
```

File: Objects/typeobject.c [hunk 4]
```diff
@@ -3666,6 +3666,32 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
...
+        for (int i=0; i<PyTuple_Size(bases); i++) {
```
Comment (erlend-aasland) [inline Objects/typeobject.c]: Since we know bases is a tuple, we can use the fast APIs. Ditto below.

File: Modules/_testcapi/heaptype.c [hunk 5]
```diff
@@ -365,6 +365,21 @@ create_type_from_repeated_slots(PyObject *self, PyObject *variant_obj)
...
 
 
+
+static PyObject *
+make_immutable_type_with_base(PyObject *self, PyObject *base)
+{
+    assert(PyType_Check(base));
+    PyType_Spec ImmutableSubclass_spec = {
+        .name = "ImmutableSubclass",
+        .basicsize = ((PyTypeObject*)base)->tp_basicsize,
```
Comment (erlend-aasland) [inline Modules/_testcapi/heaptype.c]: You've probably already seen it, but this needs type cast to `int` to get rid of the compiler warning.

File: Objects/typeobject.c (hunk 3 above)
Comment (encukou) [reply Objects/typeobject.c:3674]: PyType_From* [only creates heap types](https://github.com/python/cpython/blob/07afe1b3f9c12b686b2c0bb4e3075000eea75bf7/Objects/typeobject.c#L3741).

Comment (erlend-aasland) [reply Objects/typeobject.c:3674]: Doh, of course :)

**Review by erlend-aasland — APPROVED** [review]

**Pull request merged**